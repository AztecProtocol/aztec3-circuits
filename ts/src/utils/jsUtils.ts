export function range(n: number, offset: number = 0) {
  const ret: number[] = [];
  for (let i = 0; i < n; i++) {
    ret.push(offset + i);
  }
  return ret;
}

export function assertLength(obj: any, member: string, length: number) {
  if (obj[member].length !== length) {
    throw new Error(
      `Expected ${member} to have length ${length}! Was: ${obj[member].length}`
    );
  }
}
