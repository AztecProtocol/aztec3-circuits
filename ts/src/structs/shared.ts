import { randomBytes } from "crypto";
import { serializeToBuffer } from "../wasm/serialize.js";
import { checkLength } from "./utils.js";

export class Fr {
  static SIZE_IN_BYTES = 32;

  constructor(private buffer: Buffer) {
    if (buffer.length != Fr.SIZE_IN_BYTES) {
      throw new Error(
        `Unexpected buffer size ${buffer.length} (expected ${Fr.SIZE_IN_BYTES} bytes)`
      );
    }
  }

  static random() {
    return new Fr(
      Buffer.concat([Buffer.alloc(1, 0), randomBytes(Fr.SIZE_IN_BYTES - 1)])
    );
  }

  toString() {
    return "0x" + this.buffer.toString("hex");
  }

  toBuffer() {
    return this.buffer;
  }
}

/**
 * For Ethereum addresses, which must be treated as 32 bytes.
 * @param buffer - The 20 byte ethereum address buffer.
 * @returns The 32 byte padded buffer.
 */
function pad32(buffer: Buffer) {
  // Create a 32-byte Buffer filled with zeros
  const paddedBuffer = Buffer.alloc(32);

  // Calculate the padding length
  const paddingLength = paddedBuffer.length - buffer.length;

  // Copy the original Buffer into the padded Buffer with an offset
  buffer.copy(paddedBuffer, paddingLength);
  return paddedBuffer;
}

export class EthAddress {
  static SIZE_IN_BYTES = 20;

  constructor(private buffer: Buffer) {
    if (buffer.length != EthAddress.SIZE_IN_BYTES) {
      throw new Error(
        `Unexpected buffer size ${buffer.length} (expected ${EthAddress.SIZE_IN_BYTES} bytes)`
      );
    }
  }

  static random() {
    return new Fr(randomBytes(EthAddress.SIZE_IN_BYTES));
  }

  toString() {
    return "0x" + this.buffer.toString("hex");
  }

  toBuffer() {
    return pad32(this.buffer);
  }
}

export class MembershipWitness<N extends number> {
  constructor(pathSize: N, public leafIndex: UInt32, public siblingPath: Fr[]) {
    checkLength(this.siblingPath, pathSize, "MembershipWitness.siblingPath");
  }

  toBuffer() {
    return serializeToBuffer(this.leafIndex, ...this.siblingPath);
  }
}

export class AggregationObject {
  public hasData: false = false;
  constructor(
    public p0: G1,
    public p1: G1,
    public publicInputs: Fr[],
    public proofWitnessIndices: UInt32[]
  ) {}

  toBuffer() {
    return serializeToBuffer(
      this.p0,
      this.p1,
      this.publicInputs,
      this.proofWitnessIndices,
      this.hasData
    );
  }
}

export type UInt32 = number;

// TODO: Define proper type for AztecAddress
export type AztecAddress = Fr;

// TODO: What is a Curve::G1?
export type G1 = Fr;

export type Proof = Buffer;
export type VK = Buffer;
