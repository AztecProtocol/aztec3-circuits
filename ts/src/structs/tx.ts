import { serializeToBuffer } from "../wasm/serialize.js";
import { Fr } from "./shared.js";

/**
 * Contract deployment data in a @TxContext.
 * cpp/src/aztec3/circuits/abis/contract_deployment_data.hpp
 * 
 * @todo Used the Cpp type as source of truth, which does not match the specification.
 * Spec includes constructorVkHash but does not have contract_data_hash nor constructor_hash.
 */
export class ContractDeploymentData {
  constructor (
    public contractDataHash: Fr,
    public functionTreeRoot: Fr,
    public constructorHash: Fr,
    public contractAddressSalt: Fr,
    /**
     * EthAddress of the Portal Contract
     * TODO: Change to EthAddress once there's a uint160 circuit type
     */
    public portalContractAddress: Fr,
  ) { }

  toBuffer() {
    return serializeToBuffer(
      this.contractDataHash,
      this.functionTreeRoot,
      this.constructorHash,
      this.contractAddressSalt,
      this.portalContractAddress,
    );
  }
}

/**
 * Transaction context
 * cpp/src/aztec3/circuits/abis/tx_context.hpp
 */
export class TxContext {
  constructor (
    public isFeePaymentTx: false,
    public isRebatePaymentTx: false,
    public isContractDeployment: true,
    public contractDeploymentData: ContractDeploymentData,
    public referenceBlockNumber: Fr,
  ) {}

  toBuffer() {
    return serializeToBuffer(
      this.isFeePaymentTx,
      this.isRebatePaymentTx,
      this.isContractDeployment,
      this.contractDeploymentData,
      this.referenceBlockNumber,
    );
  }
}