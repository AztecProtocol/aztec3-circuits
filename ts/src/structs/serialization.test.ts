import { assertLength, range } from "../utils/jsUtils.js";
import { CircuitsWasm } from "../wasm/circuits_wasm.js";
import {
  numToUInt32LE,
  serializeToBuffer,
  uint8ArrayToNum,
} from "../wasm/serialize.js";
import {
  ARGS_LENGTH,
  L1_MSG_STACK_LENGTH,
  NEW_COMMITMENTS_LENGTH,
  NEW_NULLIFIERS_LENGTH,
  PRIVATE_CALL_STACK_LENGTH,
  PUBLIC_CALL_STACK_LENGTH,
  RETURN_VALUES_LENGTH,
} from "./constants.js";
import { AztecAddress, EthAddress, Fr } from "./shared.js";
import { ContractDeploymentData, TxContext } from "./tx.js";

/**
 * Call context.
 * @see abis/call_context.hpp
 */
export class CallContext {
  constructor(
    public msgSender: AztecAddress,
    public storageContractAddress: AztecAddress,
    public portalContractAddress: EthAddress,
    public isDelegateCall: boolean,
    public isStaticCall: boolean,
    public isContractDeployment: boolean
  ) {}
  toBuffer(): Buffer {
    return serializeToBuffer(
      this.msgSender,
      this.storageContractAddress,
      this.portalContractAddress,
      this.isDelegateCall,
      this.isStaticCall,
      this.isContractDeployment
    );
  }
}

export class PrivateCircuitPublicInputs {
  constructor(
    public callContext: CallContext,
    public args: Fr[],
    public returnValues: Fr[],
    public newCommitments: Fr[],
    public newNullifiers: Fr[],
    public privateCallStack: Fr[],
    public publicCallStack: Fr[],
    public l1MsgStack: Fr[],
    public historicPrivateDataTreeRoot: Fr
  ) {
    assertLength(this, "args", ARGS_LENGTH);
    assertLength(this, "returnValues", RETURN_VALUES_LENGTH);
    assertLength(this, "newCommitments", NEW_COMMITMENTS_LENGTH);
    assertLength(this, "newNullifiers", NEW_NULLIFIERS_LENGTH);
    assertLength(this, "privateCallStack", PRIVATE_CALL_STACK_LENGTH);
    assertLength(this, "publicCallStack", PUBLIC_CALL_STACK_LENGTH);
    assertLength(this, "l1MsgStack", L1_MSG_STACK_LENGTH);
  }
  toBuffer(): Buffer {
    return serializeToBuffer(
      this.callContext,
      this.args,
      this.returnValues,
      this.newCommitments,
      this.newNullifiers,
      this.privateCallStack,
      this.publicCallStack,
      this.l1MsgStack,
      this.historicPrivateDataTreeRoot
    );
  }
}

function fr(n: number) {
  return new Fr(numToUInt32LE(n));
}

function privateCircuitPublicInputs() {
  return new PrivateCircuitPublicInputs(
    new CallContext(
      numToUInt32LE(1),
      numToUInt32LE(2),
      new EthAddress(numToUInt32LE(3)),
      true,
      true,
      true
    ),
    range(ARGS_LENGTH).map(fr),
    range(RETURN_VALUES_LENGTH).map(fr),
    range(NEW_COMMITMENTS_LENGTH).map(fr),
    range(NEW_NULLIFIERS_LENGTH).map(fr),
    range(PRIVATE_CALL_STACK_LENGTH).map(fr),
    range(PUBLIC_CALL_STACK_LENGTH).map(fr),
    range(L1_MSG_STACK_LENGTH).map(fr),
    new Fr(numToUInt32LE(1))
  );
}

function txContext() {
  const deploymentData = new ContractDeploymentData(
    new Fr(numToUInt32LE(1)),
    new Fr(numToUInt32LE(2)),
    new Fr(numToUInt32LE(3)),
    new Fr(numToUInt32LE(4)),
    new Fr(numToUInt32LE(5))
  );
  return new TxContext(false, false, true, deploymentData);
}

describe("basic struct serialization", () => {
  const wasm: CircuitsWasm = new CircuitsWasm();
  beforeAll(async () => {
    await wasm.init();
  });
  for (const { name, object, method } of [
    {
      name: "PrivateCircuitPublicInputs",
      object: privateCircuitPublicInputs(),
      method: "abis__test_roundtrip_serialize_private_circuits_public_inputs",
    },
    {
      name: "TxContext",
      object: txContext(),
      method: "abis__test_roundtrip_serialize_tx_context",
    },
  ]) {
    it(`serializes a ${name} and calls trivial C++ code`, () => {
      const inputBuf = object.toBuffer();
      const inputBufPtr = wasm.call("bbmalloc", inputBuf.length);
      wasm.writeMemory(inputBufPtr, inputBuf);
      const outputBufSizePtr = wasm.call("bbmalloc", 4);
      // Get a string version of our object. As a quick and dirty test,
      // we compare a snapshot of its string form to its previous form.
      const outputBufPtr = wasm.call(method, inputBufPtr, outputBufSizePtr);
      // Read the size pointer
      const outputBufSize = uint8ArrayToNum(
        wasm.getMemorySlice(outputBufSizePtr, outputBufSizePtr + 4)
      );
      const outputBuf = wasm.getMemorySlice(
        outputBufPtr,
        outputBufPtr + outputBufSize
      );
      const outputStr = Buffer.from(outputBuf).toString("utf-8");
      expect(outputStr).toMatchSnapshot();
      // Free memory
      wasm.call("bbfree", outputBufPtr);
      wasm.call("bbfree", outputBufSizePtr);
      wasm.call("bbfree", inputBufPtr);
    });
  }
});
