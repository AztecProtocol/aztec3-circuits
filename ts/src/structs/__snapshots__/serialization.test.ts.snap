// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`basic struct serialization serializes a PrivateCircuitPublicInputs and calls trivial C++ code 1`] = `
"call_context: msg_sender: 0x1
storage_contract_address: 0x2
tx_origin: 0x3
is_delegate_call: 1
is_static_call: 1
is_contract_deployment: 1

args: [ 0x1 0x2 0x3 0x4 0x5 0x6 0x7 0x8 ]
return_values: [ 0x1 0x2 0x3 0x4 ]
emitted_events: [ 0x1 0x2 0x3 0x4 ]
new_commitments: [ 0x1 0x2 0x3 0x4 ]
new_nullifiers: [ 0x1 0x2 0x3 0x4 ]
private_call_stack: [ 0x1 0x2 0x3 0x4 ]
public_call_stack: [ 0x1 0x2 0x1 0x31000000c8cd1200000000000000000000000000000000 ]
l1_msg_stack: [ 0x612d020000000000000000000000000000000000000000 0x412d020000000000000000000000000000000000000000 ]
old_private_data_tree_root: 0x212d020000000000000000000000000000000000000000
old_nullifier_tree_root: 0x0
old_nullifier_tree_root: 0x0
contract_deployment_data: constructor_vk_hash: 0x0
function_tree_root: 0x0
contract_address_salt: 0x0
portal_contract_address: 0x0

"
`;

exports[`basic struct serialization serializes a TxContext and calls trivial C++ code 1`] = `
"is_fee_payment_tx: 0
is_rebate_payment_tx: 0
is_contract_deployment_tx: 1
contract_deployment_data: 
constructor_vk_hash: 0x1
function_tree_root: 0x2
contract_address_salt: 0x3
portal_contract_address: 0x4
"
`;
